File: src/hyperion_mixed/script/hyperion_talker.py
#!/usr/bin/env python3
# note: Hyperion is programmed under python3
import rospy
import numpy as np
from std_msgs.msg import Float32MultiArray
from std_msgs.msg import MultiArrayDimension
from omni_msgs.msg import OmniFeedback


from hyperion_command import Hcommand


def hyperion_force(address, ref_wavelen):
    '''
    return force feedback np.array([[f_x], [f_y]])
    Calculation detail is from a Chinese essay located in "/hyperion_mixed"
    '''
    try:
        # network configuration of hyperion device (we may not need this)
        # h.connect_config(current_address)

        # TODO: real test
        # retrieve hyperion data
        # get peaks and wavelengths from hyperion in 1*3 array
        '''
        wave_diff = h.getpeaks(address, ref_wavelen)
        # Calculation referenced from essay
        lbd_m = np.sum(wave_diff) / 3
        lbd_1 = wave_diff[0] - lbd_m
        lbd_2 = wave_diff[1] - lbd_m
        lbd_3 = wave_diff[2] - lbd_m

        # matrix calculation
        k_ = np.array([[ 37.47996088,   8.99352363, -46.47348451],
                        [ 27.42708426, -43.05784321,  15.63075895]])
        # force: 2*1 matrix
        f = np.dot(k_, np.array([[lbd_1], [lbd_2], [lbd_3]]))
        return f
        '''
        pass
    except:
        # print out warning message
        print("\033[1;31;45m[WARN] Hyperion get peaks failed\033[0m")
        print("\033[1;31;45m[WARN] Stopping Hyperion script\033[0m")
        sys.exit()


# calculate reference wavelength during initialization
def wavelength_calibrate(address):
    ref_wavelen = 0
    for i in range(20):
        test = h.getpeaks()
        ref_wavelen += np.sum(test)
    return(ref_wavelen / 20 + 1550)


def talker(pub_omni):

    while not rospy.is_shutdown():
        # TODO: uncomment for real test
        # f = hyperion_force()
        # f_x = f[0][0]
        # f_y = f[1][0]
        force_data = OmniFeedback()
        # force_data.force.x = f_x
        # force_data.force.y = f_y
        force_data.force.z = 0

        rospy.loginfo("python: force_feedback_data sent")
        # rospy.loginfo(_hyperion_data) # for test only
        pub_omni.publish(force_data)

        rate.sleep()


if __name__ == '__main__':
    # TODO: test publishing force to omni phantom; OmniFeedback.msg
    pub_omni = rospy.Publisher('phantom/force_feedback', OmniFeedback, queue_size=10)
    rospy.init_node('Hyperion_talker', anonymous=False)
    rate = rospy.Rate(11) # 11hz

    # TODO: cofirm hyperion address and revise eye_op_robot.launch
    # retrieve parameter from .launch file
    current_address = rospy.get_param("~Hyperion_address")

    # hyperion connection
    h = Hcommand(current_address)
    # reference wavelength calibration
    # TODO: uncomment for real test
    # ref_wavelen = wavelength_calibrate(current_address)
    # TODO: test retrieve and process hyperion data
    # f_x, f_y = hyperion_force(current_address, ref_wavelen)

    try:
        talker(pub_omni)
    except rospy.ROSInterruptException:
        rospy.logerr("hyperion_data error")
        pass

File: src/hyperion_mixed/script/hyperion_talker.cpp
#include "ros/ros.h"
#include "std_msgs/Float32MultiArray.h"
#include "../src/Hyperion_CPP_API/hLibrary.h"
#include "../src/hyperion_get_peaks.h"

#include <sstream>
#include <iostream>
#include <fstream>
#include <algorithm>
#include <time.h>
#include <unistd.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <ctype.h>
#include <vector>


void hyperion_talker(ros::NodeHandle n)
{


  ros::Publisher chatter_pub = n.advertise<std_msgs::Float32MultiArray>("Hyperion_data", 1000);

  ros::Rate loop_rate(10);

  /**
   * A count of how many messages we have sent. This is used to create
   * a unique string for each message.
   */
  int count = 0;
  while (ros::ok())
  {
    /**
     * This is a message object. You stuff it with data, and then publish it.
     */
    std_msgs::Float32MultiArray msg;

    std::vector<float> msg_data={0,0,0,1,2,3};
    msg.data = msg_data;

    ROS_INFO("%s", "hyperion data sent");

    chatter_pub.publish(msg);

    ros::spinOnce();

    loop_rate.sleep();
    ++count;
  }
}


void hyperion_commu(ros::NodeHandle nh, std::vector<double> allPeaks, std::vector<double> wavelength)
{
  // retrieve param from .launch file
  std::string current_address;
  nh.getParam("Hyperion_address", current_address);

  // use class to comunicate with hyperion
  // hyperion_operation test;
  try
  {
    ;
  }
  catch(...)
  {
    std::cout << "\033[31m [WARN] Hyperion connection failed \033[0m""" << std::endl;
    exit(0);
  }
  // test.test_print();
  // test.networkconfiguration();
  // test.networkconfiguration(current_address, "hyperion")
  // test.get_peaks(allPeaks, wavelength)

}


int main(int argc, char **argv)
{
  // ROS node initialize
  ros::init(argc, argv, "Hyperion_talker");
  ros::NodeHandle n;
  ros::NodeHandle nh("~"); // for private param in .launch

  // connect to hyperion and get data
  std::vector<double> allPeaks, wavelength;
  hyperion_commu(nh, allPeaks, wavelength);

  // ROS talker that publish data
  hyperion_talker(n);

  return 0;
}
